crates/re_build_info/src/crate_version.rs

```diff
@@ -101,99 +198,162 @@ impl CrateVersion {
+    ///
+    /// See [`CrateVersion`] for more information.
+    pub const fn try_parse(version_string: &str) -> Result<Self, &'static str> {
         // Note that this is a const function, which means we are extremely limited in what we can do!
 
-        const fn parse_u8(s: &[u8], begin: usize, end: usize) -> u8 {
-            assert!(begin < end);
-            assert!(
-                s[begin] != b'0' || begin + 1 == end,
-                "multi-digit number cannot start with zero"
-            );
-
-            let mut num = 0u64;
-            let mut i = begin;
-
-            while i < end {
-                let c = s[i];
-                assert!(
-                    b'0' <= c && c <= b'9',
-                    "Unexpected non-digit in version string"
-                );
-                let digit = c - b'0';
-                num = num * 10 + digit as u64;
-                assert!(num <= MAX_NUM as _, "Too large number in rust version");
-                i += 1;
+        const fn maybe(s: &[u8], c: u8) -> (bool, &[u8]) {
+            if !s.is_empty() && s[0] == c {
+                (true, slice!(s, 1, ..))
+            } else {
+                (false, s)
             }
-            assert!(num <= u8::MAX as u64);
-            num as _
         }
 
-        let s = version_string.as_bytes();
+        const fn maybe_token<'a>(s: &'a [u8], token: &[u8]) -> (bool, &'a [u8]) {
+            if s.len() < token.len() {
+                return (false, s);
+            }
 
-        let mut i = 0;
-        while s[i] != b'.' {
-            i += 1;
+            let (left, right) = split_at(s, token.len());
+            if equals(left, token) {
+                (true, right)
+            } else {
+                (false, s)
+            }
         }
-        let major = parse_u8(s, 0, i);
 
-        i += 1;
-        let minor_start = i;
-        while s[i] != b'.' {
-            i += 1;
+        macro_rules! eat {
+            ($s:ident, $c:expr, $msg:literal) => {{
+                if $s.is_empty() || $s[0] != $c {
+                    return Err($msg);
+                }
+                slice!($s, 1, ..)
+            }};
         }
-        let minor = parse_u8(s, minor_start, i);
 
-        i += 1;
-        let patch_start = i;
-        while i < s.len() && s[i] != b'-' && s[i] != b'+' {
-            i += 1;
+        macro_rules! eat_token {
+            ($s:ident, $token:expr, $msg:literal) => {{
+                let token = $token;
+                if token.len() > $s.len() {
+                    return Err($msg);
+                }
+                let (left, right) = split_at($s, token.len());
+                if !equals(left, token) {
+                    return Err($msg);
+                }
+                right
+            }};
         }
-        let patch = parse_u8(s, patch_start, i);
 
-        if i == s.len() {
-            return Self::new(major, minor, patch);
+        macro_rules! eat_u8 {
```

Comment:
don't quite understand why this one couldn't be a local method instead, can you point it out for me?

Reply:
`?` doesn't work in `const` and it seemed simpler to turn these into macros, so that I can return the errors directly. Another option would be to use a custom implementation of `try!` that doesn't call `From::from` on the error, but I think the actual parsing code ends up a bit cleaner this way.